components/ai_chat/browser/ai_chat_tab_helper.h

```diff
@@ -0,0 +1,124 @@
+  };
+
+  AIChatTabHelper(const AIChatTabHelper&) = delete;
+  AIChatTabHelper& operator=(const AIChatTabHelper&) = delete;
+  ~AIChatTabHelper() override;
+
+  const std::vector<mojom::ConversationTurn>& GetConversationHistory();
+  // Whether the UI for this conversation is open or not. Determines
+  // whether content is retrieved and queries are sent for the conversation
+  // when the page changes.
+  void OnConversationActiveChanged(bool is_conversation_active);
+  void AddToConversationHistory(const mojom::ConversationTurn& turn);
+  void UpdateOrCreateLastAssistantEntry(const std::string& text);
+  void MakeAPIRequestWithConversationHistoryUpdate(
+      const mojom::ConversationTurn& turn);
+  bool IsRequestInProgress();
+  void AddObserver(Observer* observer);
+  void RemoveObserver(Observer* observer);
+  // On-demand request to fetch questions related to the content. If no content
+  // is available for the current page, or if questions
+  // are already generated, nothing will happen.
+  void GenerateQuestions();
+  std::vector<std::string> GetSuggestedQuestions(bool& can_generate,
+                                                 bool& auto_generate);
+
+ private:
+  friend class content::WebContentsUserData<AIChatTabHelper>;
+
+  explicit AIChatTabHelper(content::WebContents* web_contents);
+
+  bool HasUserOptedIn();
+  std::string GetConversationHistoryString();
+  void GeneratePageText();
+  void CleanUp();
+  void OnTabContentRetrieved(int64_t for_navigation_id,
+                             std::string contents_text,
+                             bool is_video = false);
+  void OnAPIStreamDataReceived(int64_t for_navigation_id,
+                               data_decoder::DataDecoder::ValueOrError result);
+  void OnAPIStreamDataComplete(int64_t for_navigation_id,
+                               api_request_helper::APIRequestResult result);
+  void OnAPISuggestedQuestionsResponse(
+      int64_t for_navigation_id,
+      api_request_helper::APIRequestResult result);
+  void OnSuggestedQuestionsChanged();
+
+  // content::WebContentsObserver:
+  void PrimaryPageChanged(content::Page& page) override;
+  void DocumentOnLoadCompletedInPrimaryMainFrame() override;
+  void WebContentsDestroyed() override;
+  void DidFinishNavigation(
+      content::NavigationHandle* navigation_handle) override;
+
+  raw_ptr<PrefService> pref_service_;
+  std::unique_ptr<AIChatAPI> ai_chat_api_ = nullptr;
+
+  base::ObserverList<Observer> observers_;
+
+  // TODO(nullhook): Abstract the data model
+  std::vector<mojom::ConversationTurn> chat_history_;
+  std::string article_text_;
+  bool is_conversation_active_ = false;
+  bool is_request_in_progress_ = false;
+  std::vector<std::string> suggested_questions_;
+  bool has_generated_questions_ = false;
+  bool is_video_ = false;
+  // Store the unique ID for each navigation so that
+  // we can ignore API responses for previous navigations.
+  int64_t current_navigation_id_;
```

Comment:
I noticed that we operate on the `navigation_id` in the tab helper, but I was wondering if it would be better to operate on the `content_id` instead. Since the summarization is based on the page's content, can there be multiple content for a single navigation id? 

As our AXTreeDistiller matures, should we consider extracting content not only from the main frame but also from other frames available on the page? If so, how would this work with the navigation id?

Currently, the source of content is coming from the renderer. Perhaps it would be better if the renderer could also send an AXTree token, allowing us to perform checks from the browser process to verify if the token associated with the content matches. This would be similar to how we handle the navigation id.

```
ui::AXTreeID tree_id web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
current_tree_id_token_ = tree_id.token()
```

Let me know what you think

Reply:
The distilling method is the same as master, so feel free to do that as a follow-up. Can you provide more information about `content_id`? We could also look at that as a follow-up, as `navigation_handle->GetNavigationId()` is working mostly well to fix the issue. We need to look at non-navigation content and url change (i.e. history push-state on youtube.com) and I'm intending to look at that in a follow-up.